reactive/swarm.py

```diff
@@ -168,32 +177,75 @@ def enable_client_tls():
+def prepare_default_client_credentials():
     """ Generate a downloadable package for clients to use to speak to the
     swarm cluster. """
-    if is_leader():
-        client_cert('./swarm_credentials')
-        ca('./swarm_credentials')
 
-        # Prepare the workspace
-        with chdir('./swarm_credentials'):
-            rename('client.key', 'key.pem')
-            rename('client.crt', 'cert.pem')
-            rename('ca.crt', 'ca.pem')
+    # Leverage TLSLib to copy the default cert from PKI
+    client_cert('./swarm_credentials')
+    ca('./swarm_credentials')
+
+    # Prepare the workspace
+    with chdir('./swarm_credentials'):
+        rename('client.key', 'key.pem')
+        rename('client.crt', 'cert.pem')
+        rename('ca.crt', 'ca.pem')
 
-        template_vars = {'public_address': unit_get('public-address')}
+    with open('swarm_credentials/key.pem', 'r') as fp:
+        key_contents = fp.read()
+    with open('swarm_credentials/cert.pem', 'r') as fp:
+        crt_contents = fp.read()
 
-        render('enable.sh', './swarm_credentials/enable.sh', template_vars)
+    leader_set({'client_cert': crt_contents,
+                'client_key': key_contents})
 
-        cmd = 'tar cvf swarm_credentials.tar swarm_credentials'
-        subprocess.check_call(split(cmd))
-        copyfile('swarm_credentials.tar', '/home/ubuntu/swarm_credentials.tar')
-        set_state('client.credentials.placed')
+
+@when_any('leadership.changed.client_cert', 'leadership.changed.client_key')
+@when_not('client.credentials.placed')
+def prepare_end_user_package():
+    """ Prepare the tarball package for clients to use to connet to the
+        swarm cluster using the default client credentials. """
+
+    # If we are a follower, we dont have keys and need to fetch them
+    # from leader-data, which triggered `leadership.set.client_cert`
+    # So it better be there!
+    if not path.exists('swarm_credentials'):
+        makedirs('swarm_credentials')
+        with open('swarm_credentials/key.pem', 'w+') as fp:
+            fp.write(leader_get('client_key'))
+        with open('swarm_credentials/cert.pem', 'w+') as fp:
+            fp.write(leader_get('client_cert'))
+        with open('swarm_credentials/ca.pem', 'w+') as fp:
+            fp.write(leader_get('certificate_authority'))
+
+    # Render the client package script
+    template_vars = {'public_address': unit_get('public-address')}
+    render('enable.sh', './swarm_credentials/enable.sh', template_vars)
+
+    # clear out any stale credentials package
+    if path.exists('swarm_credentials.tar'):
+        remove('swarm_credentials.tar')
+
+    # Package up the client credentials
+    cmd = 'tar cvf swarm_credentials.tar swarm_credentials'
+    subprocess.check_call(split(cmd))
+    copyfile('swarm_credentials.tar', '/home/ubuntu/swarm_credentials.tar')
+    set_state('client.credentials.placed')
+
+
+@when('leadership.is_leader')
```

Comment:
Will this not run every 5 seconds? Do you mean to make that a "changed" event/state?